utils/frame/remote-externalities/src/lib.rs

```diff
@@ -462,214 +565,342 @@ where
-			let child_kv_inner =
-				self.rpc_child_get_storage_paged(prefixed_top_key, child_keys, at).await?;
-
-			let prefixed_top_key = PrefixedStorageKey::new(prefixed_top_key.clone().0);
-			let un_prefixed = match ChildType::from_prefixed_key(&prefixed_top_key) {
-				Some((ChildType::ParentKeyId, storage_key)) => storage_key,
-				None => {
-					log::error!(target: LOG_TARGET, "invalid key: {:?}", prefixed_top_key);
-					return Err("Invalid child key")
-				},
-			};
+		// NOTE: the threading done here is the simpler, yet slightly un-elegant because we are
+		// splitting child root among threads, and it is very common for these root to have vastly
+		// different child tries underneath them, causing some threads to finish way faster than
+		// others. Certainly still better than single thread though.
+		let mut handles = vec![];
+		let client = self.as_online().transport.remote_client.clone().unwrap();
+		let at = self.as_online().at_expected();
+
+		enum Message {
+			Terminated,
+			Batch((ChildInfo, Vec<(Vec<u8>, Vec<u8>)>)),
+		}
+		let (tx, rx) = std::sync::mpsc::channel::<Message>();
+
+		for thread_child_roots in child_roots
+			.chunks(child_roots_per_thread)
+			.map(|x| x.into())
+			.collect::<Vec<Vec<_>>>()
+		{
+			let thread_client = client.clone();
+			let thread_sender = tx.clone();
+			let handle = thread::spawn(move || {
+				let rt = tokio::runtime::Runtime::new().unwrap();
+				let mut thread_child_kv = vec![];
+				for prefixed_top_key in thread_child_roots {
+					let child_keys = rt.block_on(Self::rpc_child_get_keys(
+						&thread_client,
+						&prefixed_top_key,
+						StorageKey(vec![]),
+						at,
+					))?;
+					let child_kv_inner = rt.block_on(Self::rpc_child_get_storage_paged(
+						&thread_client,
+						&prefixed_top_key,
+						child_keys,
+						at,
+					))?;
+
+					let prefixed_top_key = PrefixedStorageKey::new(prefixed_top_key.clone().0);
+					let un_prefixed = match ChildType::from_prefixed_key(&prefixed_top_key) {
+						Some((ChildType::ParentKeyId, storage_key)) => storage_key,
+						None => {
+							log::error!(target: LOG_TARGET, "invalid key: {:?}", prefixed_top_key);
+							return Err("Invalid child key")
+						},
+					};
+
+					thread_sender
+						.send(Message::Batch((
+							ChildInfo::new_default(un_prefixed),
+							child_kv_inner
+								.iter()
+								.cloned()
+								.map(|(k, v)| (k.0, v.0))
+								.collect::<Vec<_>>(),
+						)))
+						.unwrap();
+					thread_child_kv.push((ChildInfo::new_default(un_prefixed), child_kv_inner));
+				}
 
-			child_kv.push((ChildInfo::new_default(un_prefixed), child_kv_inner));
+				thread_sender.send(Message::Terminated).unwrap();
+				Ok(thread_child_kv)
+			});
+			handles.push(handle);
 		}
 
+		// first, wait until all threads send a `Terminated` message, in the meantime populate
+		// `pending_ext`.
+		let mut terminated = 0usize;
+		loop {
+			match rx.recv().unwrap() {
```

Comment:
in general we shouldn't use anything "blocking" in async fn's as `std::sync::mpsc::Receiver::recv` so would be better to use `futures::channel::mpsc` or some other "async channel"

based on your tests it's probably fine but it could fuck up the `tokio async threadpool`

Reply:
No operation is being blocked by this, but are you saying that this could slow down everything?

Reply:
`rx.recv()` will block thread and it's possible that it prevents other futures being scheduled until the next `yield point`

Reply:
in other words until a response in sent on this channel.